### PR TITLE
Fix multiple lr scheduler (warmup scheduler) & Add adaptive_patience to the lr scheduler

### PR DESCRIPTION
**Pull request by sungmanc**
### Summary

This PR introduces,
* Resolving the multi LR scheduler issue by overriding the `optimizer_step()` to enable the warmup scheduling. So, warmup logic will not be enabled by `LinearWarmupScheduler` anymore.
* Add adaptive_patience to the LRScheduler
* Add unit-tests for warmup logic

TODOs,
* Need to resolve the hard-coded index for the optimizer, and scheduler (i.e. optimizers[0]).
* For future work, we might use the callback ? to enable the warm-up scheduler, need to check the feasibility. If this way works, we could remove the `warmup_steps` and `warmup_by_epochs` at the base `OTXLitModule`.
NOTE,
* I remained the lr_scheduler_configs as list type since I got an error when I changed the type to the dictionary. I think it could be handled at the next phase since it is not important for this PR.

### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have added e2e tests for validation.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

File: tests/integration/cli/test_export_inference.py [hunk 1]
```diff
@@ -20,8 +20,8 @@ def _check_relative_metric_diff(ref: float, value: float, eps: float) -> None:
 
     avg = max(0.5 * (ref + value), 1e-9)
     diff = abs(value - ref)
-
-    assert diff / avg <= eps, f"Relative difference exceeded {eps} threshold. Absolute difference: {diff}"
+    is_negative_changed = diff / avg <= eps and value <= ref
+    assert is_negative_changed, f"Relative difference exceeded {eps} threshold. Absolute difference: {diff}"
```
Comment (harimkang) [inline tests/integration/cli/test_export_inference.py]:
```suggestion
    is_positive_changed = value > ref or diff / avg <= eps
    assert is_positive_changed, f"Relative difference exceeded {eps} threshold. Absolute difference: {diff}"
```
Isn't this right?

Comment (sungmanc) [reply tests/integration/cli/test_export_inference.py]: Thanks for the interest even though it's still draft, it will be handled at the another PR

File: src/otx/core/model/module/base.py [hunk 2]
```diff
@@ -150,6 +157,7 @@ def ensure_list(item: Any) -> list:  # noqa: ANN401
             optimizer(params=self.parameters()) if callable(optimizer) else optimizer
             for optimizer in ensure_list(self.hparams.optimizer)
         ]
+        self.init_lr = optimizers[0].param_groups[0]["lr"]
```
Comment (vinnamkim) [inline src/otx/core/model/module/base.py]:
```suggestion
        # Capture initial_lr
        for optimizer in optimizers:
             for param_group in optimizer.param_groups:
                  param_group.setdefault('initial_lr', param_group["lr"])
```

File: src/otx/engine/engine.py [hunk 3]
```diff
@@ -161,6 +161,8 @@ def train(
         metric: Metric | MetricCallable | None = None,
         run_hpo: bool = False,
         hpo_config: HpoConfig | None = None,
+        warmup_steps: int = 0,
+        warmup_by_epochs: bool = False,
```
Comment (harimkang) [inline src/otx/engine/engine.py:165]: I'm still not sure if `warmup_by_epochs` is necessary.
Anyway, could you please add a docstring for this argument?

File: src/otx/core/model/module/base.py [hunk 4]
```diff
@@ -159,10 +167,32 @@ def ensure_list(item: Any) -> list:  # noqa: ANN401
...
+
+        optimizer.step(closure=closure)
+
+        if self.warmup_by_epoch and self.trainer.current_epoch < self.warmup_steps:
+            for pg in optimizer.param_groups:
+                pg["lr"] = _scale_lr(self.trainer.current_epoch, self.warmup_steps, self.init_lr)
+
+        if not self.warmup_by_epoch and (self.trainer.global_step < self.warmup_steps):
+            for pg in optimizer.param_groups:
+                pg["lr"] = _scale_lr(self.trainer.global_step, self.warmup_steps, self.init_lr)
```
Comment (vinnamkim) [inline src/otx/core/model/module/base.py]:
```suggestion
        def _scale_lr(start_point: int, end_point: int, param_group) -> float:
            return min(1.0, float(start_point + 1) / end_point) * param_group["initial_lr"]

        if self.trainer.current_epoch < self.warmup_steps:
            lr_step = self.trainer.current_epoch if self.warmup_by_epoch else self.trainer.global_step
            for pg in optimizer.param_groups:
                pg["lr"] = _scale_lr(lr_step, self.warmup_steps, pg)

        optimizer.step(closure=closure)
```

File: src/otx/recipe/classification/multi_class_cls/otx_mobilenet_v3_large.yaml [hunk 5]
```diff
@@ -12,9 +12,6 @@ optimizer:
     weight_decay: 0.0001
 
 scheduler:
-  - class_path: otx.algo.schedulers.warmup_schedulers.LinearWarmupScheduler
-    init_args:
-      num_warmup_steps: 10
```
Comment (harimkang) [inline src/otx/recipe/classification/multi_class_cls/otx_mobilenet_v3_large.yaml:17]: Why is this model missing the `warmup_steps` setting?

File: src/otx/recipe/detection/atss_r50_fpn.yaml [hunk 6]
```diff
@@ -10,9 +10,6 @@ optimizer:
     weight_decay: 0.0
 
 scheduler:
-  - class_path: otx.algo.schedulers.warmup_schedulers.LinearWarmupScheduler
-    init_args:
-      num_warmup_steps: 3
```
Comment (harimkang) [inline src/otx/recipe/detection/atss_r50_fpn.yaml:15]: Is this model also completely devoid of warmup?

File: src/otx/recipe/detection/atss_resnext101.yaml [hunk 7]
```diff
@@ -12,9 +12,6 @@ optimizer:
     weight_decay: 0.0001
 
 scheduler:
-  - class_path: otx.algo.schedulers.warmup_schedulers.LinearWarmupScheduler
-    init_args:
-      num_warmup_steps: 3
```
Comment (harimkang) [inline src/otx/recipe/detection/atss_resnext101.yaml:17]: Same here.

File: src/otx/recipe/detection/ssd_mobilenetv2.yaml [hunk 8]
```diff
@@ -12,9 +12,6 @@ optimizer:
     weight_decay: 0.0001
 
 scheduler:
-  - class_path: otx.algo.schedulers.warmup_schedulers.LinearWarmupScheduler
-    init_args:
-      num_warmup_steps: 3
```
Comment (harimkang) [inline src/otx/recipe/detection/ssd_mobilenetv2.yaml:17]: Same here.

File: src/otx/recipe/detection/yolox_l.yaml [hunk 9]
```diff
@@ -12,9 +12,6 @@ optimizer:
     weight_decay: 0.0001
 
 scheduler:
-  - class_path: otx.algo.schedulers.warmup_schedulers.LinearWarmupScheduler
-    init_args:
-      num_warmup_steps: 3
```
Comment (harimkang) [inline src/otx/recipe/detection/yolox_l.yaml:17]: Warmup is missing here.

File: src/otx/recipe/semantic_segmentation/segnext_s.yaml [hunk 10]
```diff
@@ -37,3 +34,4 @@ callback_monitor: val/Dice
 data: ../_base_/data/mmseg_base.yaml
 overrides:
   max_epochs: 170
+  warmup_steps: 100
```
Comment (harimkang) [inline src/otx/recipe/semantic_segmentation/segnext_s.yaml:37]:
```suggestion
  warmup_steps: 20
```
Isn't this right?

File: src/otx/recipe/semantic_segmentation/segnext_t.yaml [hunk 11]
```diff
@@ -37,3 +34,4 @@ callback_monitor: val/Dice
 data: ../_base_/data/mmseg_base.yaml
 overrides:
   max_epochs: 170
+  warmup_steps: 100
```
Comment (harimkang) [inline src/otx/recipe/semantic_segmentation/segnext_t.yaml:37]: Same here.

File: src/otx/core/model/module/base.py [hunk 12]
```diff
@@ -159,10 +167,32 @@ def ensure_list(item: Any) -> list:  # noqa: ANN401
...
 
+            lr_schedulers.append(lr_scheduler_config)
         return optimizers, lr_schedulers
 
+    def optimizer_step(
+        self,
+        epoch: int,
+        batch: int,
+        optimizer: torch.optim.Optimizer,
+        closure: Callable[[], Tensor | None],
```
Comment (harimkang) [inline src/otx/core/model/module/base.py]:
```suggestion
        optimizer_closure: Callable[[], Tensor | None],
```

https://github.com/Lightning-AI/pytorch-lightning/blob/f23b3b1e7fdab1d325f79f69a28706d33144f27e/src/lightning/pytorch/core/module.py#L1270

**Review by harimkang — COMMENTED** [review]
The warmup settings seem to be different on models, is this by design? Could you please double check?

File: src/otx/core/model/module/base.py (hunk 4 above)
Comment (vinnamkim) [reply src/otx/core/model/module/base.py]: Should calling `_scale_lr()` be in front of `optimizer.step(closure=closure)` to force it to be effective in top priority?

To validate this behavior, please add integration tests for the following scenario,

With cosine LR scheduler, `warmup_steps=10`, 5 iterations per 1 epoch, and training for 10 epochs, validate the LR curve is scheduled correctly.

Comment (vinnamkim) [reply src/otx/core/model/module/base.py]: In addition, while writing this comment, I notice that the current implementation can fall into the following "what we don't want".
image
To achieve "what we want", is it better to implement `ReduceLROnPlateauWithWarmup` and make it to be called for both mode `step/epoch`?

**Comment by sungmanc** [comment]
> The warmup settings seem to be different on models, is this by design? Could you please double check?

I just following the setting from the OTX1.x and I don't think we should use the same value for all models since the characteristics could vary

**Comment by harimkang** [comment]
> > The warmup settings seem to be different on models, is this by design? Could you please double check?
> 
> I just following the setting from the OTX1.x and I don't think we should use the same value for all models since the characteristics could vary

Yes If it's the same as the setting in 1.X, you can ignore it. :)

File: src/otx/engine/engine.py (hunk 3 above)
Comment (sungmanc) [reply src/otx/engine/engine.py:165]: Done, https://github.com/openvinotoolkit/training_extensions/pull/3035/commits/fceca297d1d4bf7e892a6f75f3da4fcfde943caa

File: src/otx/recipe/classification/multi_class_cls/otx_mobilenet_v3_large.yaml (hunk 5 above)
Comment (sungmanc) [reply src/otx/recipe/classification/multi_class_cls/otx_mobilenet_v3_large.yaml:17]: Done, https://github.com/openvinotoolkit/training_extensions/pull/3035/commits/3ccd4b1946381915bc2c179da8746a23a82f9a56

File: src/otx/recipe/detection/atss_r50_fpn.yaml (hunk 6 above)
Comment (sungmanc) [reply src/otx/recipe/detection/atss_r50_fpn.yaml:15]: Done, https://github.com/openvinotoolkit/training_extensions/pull/3035/commits/3ccd4b1946381915bc2c179da8746a23a82f9a56

File: src/otx/recipe/detection/atss_resnext101.yaml (hunk 7 above)
Comment (sungmanc) [reply src/otx/recipe/detection/atss_resnext101.yaml:17]: Done, https://github.com/openvinotoolkit/training_extensions/pull/3035/commits/3ccd4b1946381915bc2c179da8746a23a82f9a56

File: src/otx/recipe/detection/ssd_mobilenetv2.yaml (hunk 8 above)
Comment (sungmanc) [reply src/otx/recipe/detection/ssd_mobilenetv2.yaml:17]: Done, https://github.com/openvinotoolkit/training_extensions/pull/3035/commits/3ccd4b1946381915bc2c179da8746a23a82f9a56

File: src/otx/recipe/detection/yolox_l.yaml (hunk 9 above)
Comment (sungmanc) [reply src/otx/recipe/detection/yolox_l.yaml:17]: Done, https://github.com/openvinotoolkit/training_extensions/pull/3035/commits/3ccd4b1946381915bc2c179da8746a23a82f9a56

File: src/otx/recipe/semantic_segmentation/segnext_s.yaml (hunk 10 above)
Comment (sungmanc) [reply src/otx/recipe/semantic_segmentation/segnext_s.yaml:37]: It comes from OTX1.x.

File: src/otx/recipe/semantic_segmentation/segnext_t.yaml (hunk 11 above)
Comment (sungmanc) [reply src/otx/recipe/semantic_segmentation/segnext_t.yaml:37]: It comes from OTX1.x.

File: src/otx/core/model/module/base.py (hunk 12 above)
Comment (sungmanc) [reply src/otx/core/model/module/base.py]: Done, https://github.com/openvinotoolkit/training_extensions/pull/3035/commits/fa86c3bd4f47fa8e8d5de6f42d3f5d662415db83

File: src/otx/core/model/module/base.py (hunk 2 above)
Comment (sungmanc) [reply src/otx/core/model/module/base.py]: Done, https://github.com/openvinotoolkit/training_extensions/pull/3035/commits/fa86c3bd4f47fa8e8d5de6f42d3f5d662415db83

File: src/otx/core/model/module/base.py (hunk 4 above)
Comment (sungmanc) [reply src/otx/core/model/module/base.py]: Done, https://github.com/openvinotoolkit/training_extensions/pull/3035/commits/fa86c3bd4f47fa8e8d5de6f42d3f5d662415db83
Add test https://github.com/openvinotoolkit/training_extensions/pull/3035/commits/65a7f91f6393c0103d5557672a1af8fc0e901787

**Comment by sungmanc** [comment]
I also manually checked the warm_up works well. 
image

**Comment by sungmanc** [comment]
Close this PR, this PR(https://github.com/openvinotoolkit/training_extensions/pull/3056) will handle the warmup scheduler issue.
